packages/amplify_datastore/ios/Classes/types/model/FlutterSerializedModel.swift

```diff
@@ -156,37 +155,74 @@ struct FlutterSerializedModel: Model, JSONValueHolder {
         }
     }
 
-    private func generateSerializedData(modelSchema: ModelSchema) -> [String: Any]{
-        
+    private static func getModelSchema(flutterModelRegistration: FlutterModels, modelName: String) throws -> ModelSchema {
+        do {
+            return try FlutterDataStoreRequestUtils.getModelSchema(modelSchemas: flutterModelRegistration.modelSchemas, modelName: modelName)
+        } catch let error {
+            print("model \(modelName) is not registered.")
+            throw error
+        }
+    }
+
+    private static func generateSerializedJsonData(values: [String: JSONValue], flutterModelRegistration: FlutterModels, modelName: String) throws -> [String: Any] {
```

Comment:
> In order to persist modelName in each resolved nested model (see below example), flutterModelRegistration object is needed to pass into functions that need access to it

Is this line from the PR description referring to these changes? is iOS not passing the model type back for the nested models? I may be wrong, but I believe Android is (with the updates that were recently merged in amplify-android). Associating the nested model with a model type seems like logic we would want done on the native side if possible.

Reply:
Given the update [to the test here](https://github.com/aws-amplify/amplify-android/pull/1390/files#diff-b7e5dd19e63c6d95721c95ab2d982fb2da430b6c29b190ee7a6fc2125c502edbR169) to add the modelSchema to the SerializedModel builder, I am fairly certain this data is being returned from amplify-android now.

A few reasons for wanting this on the native side:
- I haven't looked at the code in iOS, but I would guess that this is actually less complex and error prone to do on the native side. Getting the correct model type was pretty simple in amplify-android. It is essentially [these three lines](https://github.com/aws-amplify/amplify-android/pull/1390/files#diff-4a19dd8ac30d07afb369627a6e8c6d7018ff8cdd35a84c55f4b830c91e920aefR829-R831) of code. 
- Cleaner boundaries and less coupling between amplify-ios and amplify-flutter. With the logic to get the model name in amplify-flutter, the logic for serialized models is sort of split between the two libraries without a clean division.
- Consistency between amplify-iOS and amplify-Android

Reply:
> Consistency between amplify-iOS and amplify-Android

Good point, let me take a look.

Reply:
Looked amplify-ios [`StorageEngineAdapter`](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/SQLite/StorageEngineAdapter%2BSQLite.swift#L236) and [SQL statement + result conversion](https://github.com/aws-amplify/amplify-ios/blob/c622147b17b198cdc9d15a4ff049d7aa520139e1/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/SQLite/Statement%2BModel.swift#L71-L109) implemenation, also compared it to amplify-android. There seem to be a major difference between the two platform - amplify-ios doesn't have an equivalent of `SerializedModel` as in amplify-android.

In amplify-ios, it extracts fields from sql result, and then reads Flutter passed `modelSchemas` to convert fields into desired types, and ultimately it constructs a dictionary with nested values. This dictionary then gets decoded into `JSONValue`s by [`FlutterSerializedModel`](https://github.com/aws-amplify/amplify-flutter/blob/f67efea8d19b0ce4635bd8161363d730e1023358/packages/amplify_datastore/ios/Classes/types/model/FlutterSerializedModel.swift#L30-L45) We can see here `FlutterSerializedModel` in amplify-ios expects a json like payload which is completely different from amplify-android expecting a `SerializedModel`.

Given the level of difference, I think align amplify-ios with Android is not an easy task that can be included with this PR, and it may be the call of amplify-ios maintainer considering there are many differences between the two libraries.

In addition, I found `ModelRegistry` provides similar functionality as `ModelSchemaRegistry`, I will see if I can use this class instead of injecting `flutterModelRegistration` in multiple method functions.

Reply:
> and ultimately it constructs a dictionary with nested values

From that description, the iOS implementation doesn't sound that different from Android. A SerializedModel in amplify-android is just an object that wraps a dictionary of data, a model ID, and schema info. It sounds like iOS is returning the dictionary of data and the model ID, but not any data about the schema.

I will take a look at the iOS code tomorrow. Maybe the changes are more involved than I am thinking.

Reply:
Yes the concepts are the same :D the major difference I meant is there is no middle layer (`SerializedModel`) to hold model schema info for the nested models. That's the reason I added `modelSchema` look up within `FlutterSerializedModel`. Originally, `FlutterSerializedModel` knows only the schema (passed in as a parameter of the `toMap` function) of the root model. 

Thanks for looking!

Reply:
Got it. This definitely doesn't have to block this PR, but I think is at least worth more discussion. I think it will become difficult to maintain long term with the logic split up across amplify-ios and amplify-flutter, and the two native platforms returning different things. Maybe the decision to not create a SerializedModel in iOS was intentional though.